### PR TITLE
[Enhanced Switch][Null] Missing Null pointer access warning with total/unconditional patterns in case labels

### PR DESCRIPTION
**Pull request by srikanth-sankaran**
* Fixes https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3381

### CODE REVIEW

**Comment by srikanth-sankaran** [comment]
Expecting some tests to fail

org.eclipse.jdt.core.tests.compiler.regression.NullAnnotationTests21.test_totalTypePatternDoesNotAdmitNull()
org.eclipse.jdt.core.tests.compiler.regression.NullAnnotationTests21.test_totalTypePatternNullableExpression()

but these need to be studied to see if expectation is valid

**Comment by srikanth-sankaran** [comment]
Looking at `org.eclipse.jdt.core.tests.compiler.regression.NullAnnotationTests21.test_totalTypePatternNullableExpression()` it seems the understanding is that an unconditional pattern would match null - but that is not so.

So when `case Number n0` is matched. n0 can never be null, so the warning `Null type mismatch: required \'@NonNull Number\' but the provided value is inferred as @Nullable` expected by the test looks wrong.

@stephan-herrmann - The proposed patch is still WIP, comments are welcome

**Comment by srikanth-sankaran** [comment]
I think I have remastered the "failing" tests with right expectations. @stephan-herrmann thanks for your review.

**Comment by stephan-herrmann** [comment]
> So when `case Number n0` is matched. n0 can never be null

looks like this changed between Java 18 and 19.

> I think I have remastered the "failing" tests with right expectations. @stephan-herrmann thanks for your review.

looking into it now.

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests21.java [hunk 1]
```diff
@@ -1244,4 +1238,46 @@ static Stuff match(PatternMatching pm, int v) {
...
+				1. ERROR in X.java (at line 6)
+					switch (i) {
+					        ^
+				Null pointer access: The variable i can only be null at this location
+				----------
+				2. ERROR in X.java (at line 12)
+					switch (i) {
+					        ^
+				Null pointer access: This expression of type Integer is null but requires auto-unboxing
+				----------
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests21.java:1279]: @srikanth-sankaran  do you have an idea why these are different error messages?

**Comment by stephan-herrmann** [comment]
@srikanth-sankaran  your turn again :)

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests21.java (hunk 1 above)
Comment (srikanth-sankaran) [reply org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/NullAnnotationTests21.java:1279]: > @srikanth-sankaran do you have an idea why these are different error messages?

Yes, I had looked into this. In the former case its is a type switch which doesn't call for unboxing and in the latter switch on the unboxed integer. This results in a special message in the latter case.

**Comment by srikanth-sankaran** [comment]
Your additional changes look good, Thanks @stephan-herrmann!

**Comment by srikanth-sankaran** [comment]
The first commit from here also fixes https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3319 - basically the isNullHostile method's wrong result in the presence of total patterns was inhibiting a diagnostic from being reported.

**Pull request merged**